### PR TITLE
[ARM] Switch to soft promoting half types.

### LINKED ISSUE #73805: Wrong floating point promotion during type legalisation
```c++
#include <stdio.h>

__attribute__((noinline))
float f(_Float16 a, _Float16 b) {
  a += b;
  return a;
}

int main(void) {
  printf("%f\n", f(32, 65504));
}
```
`float` is capable of representing 65536, but `_Float16` is not, and the assignment to `a` means that holding the value in a larger type is not permitted: casts and assignments are supposed to discard excess precision and range. This is supposed to print `inf` on all targets where `_Float16` is supported.

However, on ARM (testing on QEmu), I am seeing:

```sh
$ bin/clang -O0 --target=armv7a-linux-gnueabihf test.c && qemu-arm -L /usr/arm-linux-gnueabihf a.out
inf
$ bin/clang -O1 --target=armv7a-linux-gnueabihf test.c && qemu-arm -L /usr/arm-linux-gnueabihf a.out
65536.000000
```
Same as `-O1` for other optimisation levels too.

Clang initially generates:

```llvm
; Function Attrs: noinline nounwind
define dso_local float @f(half noundef %a, half noundef %b) #0 {
entry:
  %a.addr = alloca half, align 2
  %b.addr = alloca half, align 2
  store half %a, ptr %a.addr, align 2, !tbaa !5
  store half %b, ptr %b.addr, align 2, !tbaa !5
  %0 = load half, ptr %b.addr, align 2, !tbaa !5
  %ext = fpext half %0 to float
  %1 = load half, ptr %a.addr, align 2, !tbaa !5
  %conv = fpext half %1 to float
  %add = fadd float %conv, %ext
  %conv1 = fptrunc float %add to half
  store half %conv1, ptr %a.addr, align 2, !tbaa !5
  %2 = load half, ptr %a.addr, align 2, !tbaa !5
  %conv2 = fpext half %2 to float
  ret float %conv2
}
```
That is, it performs the addition in `float`, and converts the result to `half`. This is a perfectly valid way of doing this.

InstCombine then transforms that as

```diff
-*** IR Dump Before InstCombinePass on f ***
+*** IR Dump After InstCombinePass on f ***
 ; Function Attrs: noinline nounwind
 define dso_local float @f(half noundef %a, half noundef %b) local_unnamed_addr #0 {
 entry:
-  %ext = fpext half %b to float
-  %conv = fpext half %a to float
-  %add = fadd float %conv, %ext
-  %conv1 = fptrunc float %add to half
+  %conv1 = fadd half %a, %b
   %conv2 = fpext half %conv1 to float
   ret float %conv2
 }
```
Turning it into an `fadd` on type `half` directly. This is not necessarily a useful transformation when `fadd half` is not a legal operation, but it is a valid transformation that preserves the behaviour.

During type legalisation, though, because of it not being a legal operation, it is again converted back to an `fadd` on type `float`:

```diff
@@ -1,13 +1,12 @@
   t0: ch,glue = EntryToken
             t2: f32,ch = CopyFromReg t0, Register:f32 %0
           t5: i32 = bitcast t2
-          t6: i16 = truncate t5
-        t7: f16 = bitcast t6
+        t18: i32 = and t5, Constant:i32<65535>
+      t16: f32 = fp16_to_fp t18
             t4: f32,ch = CopyFromReg t0, Register:f32 %1
           t8: i32 = bitcast t4
-          t9: i16 = truncate t8
-        t10: f16 = bitcast t9
-      t11: f16 = fadd t7, t10
-    t12: f32 = fp_extend t11
-  t14: ch,glue = CopyToReg t0, Register:f32 $s0, t12
+        t21: i32 = and t8, Constant:i32<65535>
+      t19: f32 = fp16_to_fp t21
+    t20: f32 = fadd t16, t19
+  t14: ch,glue = CopyToReg t0, Register:f32 $s0, t20
   t15: ch = ARMISD::RET_GLUE t14, Register:f32 $s0, t14:1
```

However, in this promotion back to `float`, the truncation is lost, hence the changed result. There seem to be no relevant flags (beyond general "don't optimise") to control this.

I initially thought this was another instance of #44218 but this seems like a completely different part of the compiler where things go wrong. But attempts to fix #44218 may, depending on how they are done, result in this issue.

### PR DESCRIPTION
**Pull request by hvdijk**
The traditional promotion is known to generate wrong code.

Fixes #73805.

### CODE REVIEW

**Comment by efriedma-quic** [comment]
I'm nervous about making type legalization disagree with getTypeToTransformTo().  For all other cases, it does agree, and some legalization code assumes it agrees.  Particularly for operations with f16 operands and non-f16 results.  The fact that this works at all seems to lean on very subtle invariants on conversions between f16 and non-f16 values.  And really, there's no reason it needs to disagree; the reason you're making it work this way is mostly just to avoid modifying existing code.

That said, I can't find anything specific that breaks.

**Comment by hvdijk** [comment]
> I'm nervous about making type legalization disagree with getTypeToTransformTo(). For all other cases, it does agree, and some legalization code assumes it agrees.

Note that the type legalization here is using the same type legalization that x86 and RISC-V have already been using: if `f16` is not a legal type, `TransformToType[MVT::f16]` always gets set to `MVT::f32`, yet despite that, x86 and RISC-V already use soft-promotion which transforms them to `i16`. This commit (or the previous one) does not introduce that as a new thing in LLVM, it uses what is already there. You may be right that there is code that makes assumptions that do not hold under soft-promotion, but if there is, I would expect it to show up on x86 and RISC-V as well.

File: llvm/test/CodeGen/ARM/arm-half-promote.ll [hunk 1]
```diff
@@ -2,78 +2,113 @@
...
+; CHECK:      vmov.f32        s3, s18
+; CHECK:      vmov.f32        s4, s12
+; CHECK:      vmov.f32        s5, s20
+; CHECK:      vmov.f32        s6, s14
+; CHECK:      vmov.f32        s7, s22
+; CHECK:      vmov.f32        s9, s16
+; CHECK:      vmov.f32        s11, s18
+; CHECK:      vmov.f32        s13, s20
+; CHECK:      vmov.f32        s15, s22
+; CHECK:      vpop    {d8, d9, d10, d11}
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/ARM/arm-half-promote.ll:35]: Any idea what's happening here?  Not a big deal, but seems like something we should be able to optimize.

**Review by efriedma-quic — APPROVED** [review]
Okay, then LGTM.

File: llvm/test/CodeGen/ARM/arm-half-promote.ll (hunk 1 above)
Comment (hvdijk) [reply llvm/test/CodeGen/ARM/arm-half-promote.ll:35]: We end up during ISel with
```
Creating fp constant: t2: v8f16 = BUILD_VECTOR ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>, ConstantFP:f16<APFloat(0)>
Creating new node: t3: v8f16,v8f16 = merge_values t2, t2
Creating new node: t5: f16 = extract_vector_elt t3, Constant:i32<0>
```
and quite a lot more of those, where we should be able to see that we're extracting a constant element of a constant vector, but don't appear to have any logic anywhere yet for handling that. Or if we do, where it's not working.

Comment (hvdijk) [reply llvm/test/CodeGen/ARM/arm-half-promote.ll:35]: Looking at `DAGCombiner::visitEXTRACT_VECTOR_ELT`, the logic there only handles when the operand is a `BUILD_VECTOR` directly, not when it's wrapped in `merge_values`. That might be an easy update. Would you prefer to handle that first, or at some later time?

Comment (hvdijk) [reply llvm/test/CodeGen/ARM/arm-half-promote.ll:35]: Oh, there's two reasons we don't optimise this. Firstly, `v8f16` is not a legal type and we only optimise legal types. Secondly, null constants are special cased as getting optimised even if the vector has multiple uses, but null FP constants are not null constants. We get better codegen with
```diff
diff --git a/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp b/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
index b17724cd0720..c707de4ad94b 100644
--- a/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
@@ -22241,8 +22241,7 @@ SDValue DAGCombiner::visitEXTRACT_VECTOR_ELT(SDNode *N) {
 
   // extract_vector_elt (build_vector x, y), 1 -> y
   if (((IndexC && VecOp.getOpcode() == ISD::BUILD_VECTOR) ||
-       VecOp.getOpcode() == ISD::SPLAT_VECTOR) &&
-      TLI.isTypeLegal(VecVT)) {
+       VecOp.getOpcode() == ISD::SPLAT_VECTOR)) {
     assert((VecOp.getOpcode() != ISD::BUILD_VECTOR ||
             VecVT.isFixedLengthVector()) &&
            "BUILD_VECTOR used for scalable vectors");
@@ -22252,7 +22251,7 @@ SDValue DAGCombiner::visitEXTRACT_VECTOR_ELT(SDNode *N) {
     EVT InEltVT = Elt.getValueType();
 
     if (VecOp.hasOneUse() || TLI.aggressivelyPreferBuildVectorSources(VecVT) ||
-        isNullConstant(Elt)) {
+        isNullConstant(Elt) || isNullFPConstant(Elt)) {
       // Sometimes build_vector's scalar input types do not match result type.
       if (ScalarVT == InEltVT)
         return Elt;
```
But presumably there is a reason why things are done the way they are. Let me merge this PR as is, and see what improvements work as a followup PR.

**Pull request merged**